### DISCUSSION TITLE
Type-safe register / name props

### DISCUSSION THREAD

**Discussion by hperl** [react-hook-form/react-hook-form#1869] [Q&A] [answered]
Hi everyone,

I would like to make the TypeScript compiler check iff. the names I use in the rendering code actually correspond to properties of the form data.

I see very limited support for this already when manually calling `form.register` and using a flat object (no nested arrays and objects).

## Examples

This will give a type error on "baz", because "baz" is not a property of the FormValues type.
```ts
const C = () => {
  const form = useForm<{ foo: string; bar: string }>();
  form.register("baz");
}
```

However, this will compile without errors, because for nested types the expected type of `render` is just `string`:
```ts
const C = () => {
  const form = useForm<{ foo: string; bar: string, arr: [] }>();
  form.register("baz");
}
```

And sadly, it also does not typecheck when used in render:
```ts
const C = () => {
  const form = useForm<{ foo: string; bar: string, arr: [] }>();
  return <input name="baz" ref={form.register()} />
}
```

## Question

What are best practices / techniques to ensure that the names match the properties? It feels like the TypeScript compiler has all the information to help, but I haven't connected the dots yet.

**Comment by bluebill1049** [comment, marked as answer]
@kotarella1110 is working a new plugin for strict typed form: https://github.com/react-hook-form/typed-field maybe this is what you are after.

**Reply by aspirisen** [reply]
In this case you need to use render prop for the component that you want to show, maybe it is worth to have such in register?
